pkg/build/strategies/sti/sti.go

```diff
@@ -591,16 +590,17 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
 	go dockerpkg.StreamContainerIO(errReader, &errOutput, func(a ...interface{}) { glog.Info(a...) })
 
 	err := builder.docker.RunContainer(opts)
-	if util.IsTimeoutError(err) {
-		// Cancel waiting for source input if the container timeouts
-		wg.Done()
-	}
 	if e, ok := err.(errors.ContainerError); ok {
 		// even with deferred close above, close errReader now so we avoid data race condition on errOutput;
 		// closing will cause StreamContainerIO to exit, thus releasing the writer in the equation
 		errReader.Close()
 		return errors.NewContainerError(config.BuilderImage, e.ErrorCode, errOutput)
 	}
+	// Do not wait for source input if the container times out.
+	// FIXME: this potentially leaks a goroutine.
+	if !util.IsTimeoutError(err) {
+		wg.Wait()
+	}
```

Comment:
i think we need the if-check around the wg.Wait() because the bad scenario would be:

1) injections start (wg.Add(1) occurs)
2) call to start container times out (container never starts)
3) injections never finish because the container never started
4) wg.Done() never gets called because the injections never finish
5) wg.Wait() never completes

which is the original hang we were trying to fix.

that's not to say a rewrite couldn't improve it, but at least as it stands now i believe it's necessary.